core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleLookupJTS.java

```diff
@@ -17,110 +17,113 @@
  */
 package com.graphhopper.routing.util.spatialrules;
 
-import com.graphhopper.util.shapes.GHPoint;
+import java.util.*;
+
 import org.locationtech.jts.geom.*;
 import org.locationtech.jts.index.strtree.STRtree;
 
-import java.util.*;
-
 /**
  * @author Thomas Butz
  */
 public class SpatialRuleLookupJTS implements SpatialRuleLookup {
-
+    
+    private static final Comparator<SpatialRule> RULE_COMP = new Comparator<SpatialRule>() {
+
+        @Override
+        public int compare(SpatialRule o1, SpatialRule o2) {
+            int comp = Integer.compare(o1.getPriority(), o2.getPriority());
+            if (comp != 0) {
+                return comp;
+            }
+            
+            return o1.getId().compareTo(o2.getId());
+        }
+    };
+    
     private final GeometryFactory geometryFactory = new GeometryFactory();
     private final List<SpatialRule> rules;
     private final Envelope maxBounds;
     private final STRtree index;
-
+    
     public SpatialRuleLookupJTS(List<SpatialRule> spatialRules, Envelope maxBounds) {
         this.index = new STRtree();
-
+        
         Map<Polygon, SpatialRuleContainer> containerMap = new HashMap<>();
+        List<SpatialRule> registeredRules = new ArrayList<>();
         Set<String> ruleIDs = new HashSet<>();
         for (SpatialRule rule : spatialRules) {
             if (rule == null)
                 throw new IllegalArgumentException("rule cannot be null");
-
-            if (rule.equals(SpatialRule.EMPTY))
-                throw new IllegalArgumentException("rule cannot be EMPTY");
-
+            
             if (!ruleIDs.add(rule.getId()))
                 throw new IllegalArgumentException("Duplicate rule ID: \"" + rule.getId() + "\"");
-
+            
+            boolean registered = false;
             for (Polygon border : rule.getBorders()) {
                 Envelope borderEnvelope = border.getEnvelopeInternal();
                 if (!maxBounds.intersects(borderEnvelope)) {
                     continue;
                 }
-
+                
                 SpatialRuleContainer container = containerMap.get(border);
                 if (container == null) {
                     container = new SpatialRuleContainer(border);
                     containerMap.put(border, container);
                     index.insert(borderEnvelope, container);
                 }
                 container.addRule(rule);
+                registered = true;
+            }
+            
+            if (registered) {
+                registeredRules.add(rule);
             }
         }
 
         index.build();
 
-        this.rules = new ArrayList<>();
-        rules.add(SpatialRule.EMPTY);
-        rules.addAll(spatialRules);
+        Collections.sort(registeredRules, RULE_COMP); // keep the spatial id stable
```

Comment:
Why is sorting necessary? 

See https://github.com/graphhopper/graphhopper/pull/1911/files#diff-e313449003aba3db12b92952cd486e32L85-L87 where I had to remove it to avoid problems.

Reply:
I think we can drop it here as we're already passing in a sorted collection.